server-common/src/main/java/org/apache/kafka/server/common/Features.java

```diff
@@ -16,72 +16,135 @@
+     */
+    TEST_VERSION("test.feature.version", TestFeatureVersion.values(), TestFeatureVersion::fromFeatureLevel, false);
 
-public final class Features {
-    private final MetadataVersion version;
-    private final Map<String, Short> finalizedFeatures;
-    private final long finalizedFeaturesEpoch;
+    public static final Features[] FEATURES;
+    public static final List<Features> PRODUCTION_FEATURES;
+    private final String name;
+    private final FeatureVersion[] features;
+    private final CreateMethod createFeatureVersionMethod;
+    private final boolean usedInProduction;
 
-    public static Features fromKRaftVersion(MetadataVersion version) {
-        return new Features(version, Collections.emptyMap(), -1, true);
+    Features(String name,
+             FeatureVersion[] features,
+             CreateMethod createMethod,
+             boolean usedInProduction) {
+        this.name = name;
+        this.features = features;
+        this.createFeatureVersionMethod = createMethod;
+        this.usedInProduction = usedInProduction;
     }
 
-    public Features(
-        MetadataVersion version,
-        Map<String, Short> finalizedFeatures,
-        long finalizedFeaturesEpoch,
-        boolean kraftMode
-    ) {
-        this.version = version;
-        this.finalizedFeatures = new HashMap<>(finalizedFeatures);
-        this.finalizedFeaturesEpoch = finalizedFeaturesEpoch;
-        // In KRaft mode, we always include the metadata version in the features map.
-        // In ZK mode, we never include it.
-        if (kraftMode) {
-            this.finalizedFeatures.put(FEATURE_NAME, version.featureLevel());
-        } else {
-            this.finalizedFeatures.remove(FEATURE_NAME);
-        }
+    static {
+        Features[] enumValues = Features.values();
+        FEATURES = Arrays.copyOf(enumValues, enumValues.length);
+
+        PRODUCTION_FEATURES = Arrays.stream(FEATURES).filter(feature ->
+                feature.usedInProduction).collect(Collectors.toList());
     }
 
-    public MetadataVersion metadataVersion() {
-        return version;
+    public String featureName() {
+        return name;
     }
 
-    public Map<String, Short> finalizedFeatures() {
-        return finalizedFeatures;
+    public FeatureVersion[] features() {
+        return features;
     }
 
-    public long finalizedFeaturesEpoch() {
-        return finalizedFeaturesEpoch;
+    /**
+     * Creates a FeatureVersion from a given name and level with the correct feature object underneath.
+     *
+     * @param level   the level of the feature
+     * @returns       the FeatureVersionUtils.FeatureVersion for the feature the enum is based on.
+     * @throws        IllegalArgumentException if the feature name is not valid (not implemented for this method)
+     */
+    public FeatureVersion fromFeatureLevel(short level) {
+        return createFeatureVersionMethod.fromFeatureLevel(level);
```

Comment:
Can we just iterate over `features` here and if `features.featureLevel() == level` then return, if not found throw?

Reply:
Iterate through all the features and all the levels of each feature? We could I suppose, but this could get slow if we have a lot of different features and levels.